tsdb/exemplar_test.go

```diff
@@ -119,11 +120,521 @@ func TestAddExemplar(t *testing.T) {
+				{Labels: series1, Value: 0.4, Ts: 2},
+				{Labels: series1, Value: 0.2, Ts: 3},
+				{Labels: series1, Value: 0.3, Ts: 4},
+			},
+		},
+		{
+			name: "insert out of the OOO window",
+			exemplars: []exemplar.Exemplar{
+				{Labels: series1, Value: 0.1, Ts: 200},
+				{Labels: series1, Value: 0.2, Ts: 1},
+			},
+			wantError: storage.ErrOutOfOrderExemplar,
+		},
+		{
+			name: "insert multiple series",
+			exemplars: []exemplar.Exemplar{
+				{Labels: series1, Value: 0.1, Ts: 1},
+				{Labels: series1, Value: 0.2, Ts: 3},
+				{Labels: series2, Value: 0.3, Ts: 4},
+			},
+			matcher: series1Matcher,
+			wantExemplars: []exemplar.Exemplar{
+				{Labels: series1, Value: 0.1, Ts: 1},
+				{Labels: series1, Value: 0.2, Ts: 3},
+			},
+		},
+		{
+			name: "insert multiple series with overflow",
+			exemplars: []exemplar.Exemplar{
+				{Labels: series2, Value: 0.1, Ts: 1},
+				{Labels: series2, Value: 0.2, Ts: 2},
+				{Labels: series2, Value: 0.3, Ts: 3},
+				{Labels: series1, Value: 0.4, Ts: 4},
+			},
+			matcher: series2Matcher,
+			wantExemplars: []exemplar.Exemplar{
+				{Labels: series2, Value: 0.2, Ts: 2},
+				{Labels: series2, Value: 0.3, Ts: 3},
+			},
+		},
+		{
+			name: "series1 overflows series2 out-of-order",
+			exemplars: []exemplar.Exemplar{
+				{Labels: series2, Value: 0.1, Ts: 3},
+				{Labels: series2, Value: 0.2, Ts: 2},
+				{Labels: series2, Value: 0.3, Ts: 4},
+				{Labels: series1, Value: 0.4, Ts: 4},
+				{Labels: series1, Value: 0.5, Ts: 1},
+			},
+			matcher: series2Matcher,
+			wantExemplars: []exemplar.Exemplar{
+				{Labels: series2, Value: 0.3, Ts: 4},
+			},
+		},
+		{
+			name: "ignore duplicate exemplars",
+			exemplars: []exemplar.Exemplar{
+				{Labels: series1, Value: 0.1, Ts: 3},
+				{Labels: series1, Value: 0.1, Ts: 3},
+			},
+			matcher: series1Matcher,
+			wantExemplars: []exemplar.Exemplar{
+				{Labels: series1, Value: 0.1, Ts: 3},
+			},
+		},
+		{
+			name: "empty timestamps",
+			exemplars: []exemplar.Exemplar{
+				{Labels: series1, Value: 0.1, Ts: 0},
+				{Labels: series1, Value: 0.2, Ts: 0},
+			},
+			matcher: series1Matcher,
+			wantExemplars: []exemplar.Exemplar{
+				{Labels: series1, Value: 0.1, Ts: 0},
+				{Labels: series1, Value: 0.2, Ts: 0},
+			},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			exs, err := NewCircularExemplarStorage(3, eMetrics, 100)
+			require.NoError(t, err)
+			es := exs.(*CircularExemplarStorage)
+
+			// Add exemplars and compare tc.wantErr against the first exemplar failing.
+			var addError error
+			for i, ex := range tc.exemplars {
+				addError = es.AddExemplar(ex.Labels, ex)
+				if addError != nil {
+					break
+				}
+				if testing.Verbose() {
+					t.Logf("Buffer[%d]:\n%s", i, debugCircularBuffer(es))
+				}
+			}
+			if tc.wantError == nil {
+				require.NoError(t, addError)
+			} else {
+				require.ErrorIs(t, addError, tc.wantError)
+			}
+			if (addError != nil) != (tc.wantError != nil) {
```

Comment:
isn't the `if` above doing the same job as this one? `require.` assertions immedially abort the test, why do you have the fatal?